server/historian/packages/historian-base/src/services/redisCache.ts

```diff
@@ -44,6 +44,21 @@ export class RedisCache implements ICache {
         }
     }
 
+    public async deleteIfExists(key: string): Promise<void> {
+        const exists = await this.client.exists(this.getKey(key));
+        if (!exists) {
+            return;
+        }
+
+        const result = await this.client.del(this.getKey(key));
+        // The DEL API in Redis returns the number of keys that were removed.
+        // If the key exists and we try to delete it, we expect a result equal to 1
+        // to indicate success
+        if (result !== 1) {
+            return Promise.reject(new Error(`Unable to delete key ${this.getKey(key)} from Redis.`));
+        }
```

Comment:
Is it possible that redis failed to delete an existing key but still returned success? Or the call will fail itself. I am wondering whether we need to call exists at all?

Reply:
1 and 3 are valid and no need to check the response count. For 2, does Redis return '0' or an 'error'?

Reply:
The only return to the DEL command in Redis is a number. I could not find any information out there about a different "error" type that could be returned. So to be on the safe side and be able to tell 2 from 3, I thought it would be better to pay the price of making the "exists" call. We are checking only one key, so exists is O(1) according to Redis docs (plus the network call). Do you think we do not need to worry about this scenario, thus being able to avoid the "exists" call?

Reply:
I found some interesting information in [Redis docs](https://redis.io/topics/transactions#:~:text=Redis%20commands%20can%20fail%20only%20if%20called%20with,be%20detected%20during%20development%2C%20and%20not%20in%20production.), which made me understand that the DEL command should not fail. If a failure happens, it will be called by ioredis and the promise would be rejected, I guess. So I think we might not need the `exists` check after all

"Redis commands can fail only if called with a wrong syntax (and the problem is not detectable during the command queueing), or against keys holding the wrong data type: this means that in practical terms a failing command is the result of a programming errors, and a kind of error that is very likely to be detected during development, and not in production."